### PR TITLE
PCHR-478: Datepicker must fit into window

### PR DESCRIPTION
**Pull request by kkalamarski**
## Problem:
Sometimes datepicker wasn't fully visible.
before1
## Solution:
When datepicker is opened calculate distance from window edge. If needed change position of datepicker.
after1

### CODE REVIEW

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js [hunk 1]
```diff
@@ -16,6 +16,23 @@ define([], function () {
                  */
                 attrs.datepickerPopup = HR_settings.DATE_FORMAT;
 
+                scope.$watch('isOpen', function (value) {
```
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js]: Please move the logic to a separate function

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js [hunk 2]
```diff
@@ -16,6 +16,23 @@ define([], function () {
...
+                    if (value) {
+                        scope.popupPosition = $position.position(element);
+                        scope.popupPosition.top += scope.popupPosition.height;
+
+                        var popupWidth = 316;
```
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js]: Two things:
1) This seems like a magic constant. Do we need this exact number, or can we extrapolate it in some way? If we cannot, please move it to a "constant" with a comment as to why we need this number
2) Please always declare `var`s at the beginning of the function

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js [hunk 3]
```diff
@@ -16,6 +16,23 @@ define([], function () {
...
+
+                        var popupPosition = element.offset().left + element.width();
+                        popupPosition += popupWidth;
+                        
+                        if( popupPosition > $window.innerWidth ){
```
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js]:
```js
if (popupPosition > $window.innerWidth) {
```

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js [hunk 4]
```diff
@@ -16,22 +16,30 @@ define([], function () {
                  */
                 attrs.datepickerPopup = HR_settings.DATE_FORMAT;
 
-                scope.$watch('isOpen', function (value) {
+                /**
+                 * Calculates position of the datepicker popup.
+                 * @param value
+                 */
+                var adjustPopupPosition = function (value) {
```
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js]: You could move this function definition outside the `compile` function, to make the latter cleaner and more readable
```js
var original_link = $delegate[0].link;

function adjustPopupPosition(value) {
    // ...
}

$delegate[0].compile = function ()  {
    return function (scope, element, attrs, ngModel) {
        attrs.datepickerPopup = HR_settings.DATE_FORMAT;
        
        scope.$watch('isOpen', adjustPopupPosition);
        original_link.apply(this, arguments);
    }
};

// ...
```

Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js]: Also, please put the entire module in strict mode

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js [hunk 5]
```diff
@@ -16,22 +16,30 @@ define([], function () {
                  */
                 attrs.datepickerPopup = HR_settings.DATE_FORMAT;
 
-                scope.$watch('isOpen', function (value) {
+                /**
+                 * Calculates position of the datepicker popup.
+                 * @param value
```
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js]: Define the type of `value` and give a brief description of what it is

```js
/**
 * Calculates position of the datepicker popup.
 * @param {int} value - Lorem ipsum
 */

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js [hunk 6]
```diff
@@ -16,22 +16,30 @@ define([], function () {
...
+                 */
+                var adjustPopupPosition = function (value) {
                     if (value) {
+                        var popupPosition;
+                        var popupWidth = element.parent().find('.datepicker-popup').outerWidth();
+
+
```
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js]: Unnecessary new line

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js (hunk 4 above)
Comment (kkalamarski) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js]: I'm using `scope` and `element` vars inside this function. If I move this function outside compile closure, these variables will become undefined.

Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-popup.js]: You're right, my bad!

**Comment by AkA84** [comment]
Approved

**Pull request merged**